### PR TITLE
backupccl: always track opened writer

### LINKED ISSUE #103597: backupccl: nil pointer crash in `storage.(*Writer).open` during backup in 22.2.9
During a roachtest (#103228), two nodes crashed while a backup was taken, both due to a panic within the GCS library:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x4127782]

goroutine 309871 [running]:
panic({0x4bb3320, 0x8fb3dd0})
        GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc014aeddf8 sp=0xc014aedd38 pc=0x48d77a
runtime.panicmem(...)
        GOROOT/src/runtime/panic.go:260
runtime.sigpanic()
        GOROOT/src/runtime/signal_unix.go:835 +0x2f6 fp=0xc014aede48 sp=0xc014aeddf8 pc=0x4a4636
cloud.google.com/go/storage.(*Writer).open.func1()
        cloud.google.com/go/storage/external/com_google_cloud_go_storage/writer.go:162 +0x1a2 fp=0xc014aedfe0 sp=0xc014aede48 pc=0x4127782
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc014aedfe8 sp=0xc014aedfe0 pc=0x4c2401
created by cloud.google.com/go/storage.(*Writer).open
        cloud.google.com/go/storage/external/com_google_cloud_go_storage/writer.go:152 +0x495
```

Every node in the 4-node cluster was running 22.2.9.

Stack traces for the two nodes that crashed (n3 and n4) are attached below. Note that a very similar crash had been reported before [1], and deemed fixed by [2]. However, the issue doesn't seem to be completely solved.

### Reproduction

Running the `backup-restore/mixed-version` roachtest in #103228 with seed `-4303022106448172299` seems to reproduce this with high probability (~1h30m after test start).

[n3_stacks.txt](https://github.com/cockroachdb/cockroach/files/11509063/n3_stacks.txt)
[n4_stacks.txt](https://github.com/cockroachdb/cockroach/files/11509064/n4_stacks.txt)

Roachtest artifacts: <https://console.cloud.google.com/storage/browser/cockroach-tmp/103597/roachtest_artifacts;tab=objects?project=cockroach-shared&prefix=&forceOnObjectsSortingFiltering=false&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))>

[1] https://github.com/googleapis/google-cloud-go/issues/4167
[2] https://github.com/cockroachdb/cockroach/pull/65660 

Jira issue: CRDB-28094

### PR DESCRIPTION
**Pull request by dt**
Previously we would do things -- like wrap the opened writer with an encryption shim -- after opening the writer but before storing it in the sink. This could mean that if the sink opened a writer, but failed to save it, and was then closed it would fail to close the writer. This changes that, so that as soon as the writer is opened it is saved and then if it is later wrapped, saved again, to ensure that we cannot lose track of any successfully opened writer.

Fixes: #103597.

Release note: none.

### CODE REVIEW

**Review by stevendanna — APPROVED** [review]
Funzies.

I do wonder if we could write a unit test for this case, but I won't block on that since we have a failing test that this is fixing.

**Comment by dt** [comment]
One tricky thing about a unit test is that crash only happens when using google _cloud_ storage, which by default we'd never do during _unit_ tests.

I thought about adding my mutex-wrapped-map of open writers that I used to debug this to nodelocal or userfile so they could synthesize the same crash but not sure it is worth it?

In any case, merging the fix now to unblock @renatolabs and will keep pondering if there is a better unit test.

bors r+

**Pull request merged**
